File: record/android/src/main/java/com/llfbandit/record/MediaRecorder.java
package com.llfbandit.record;

import android.content.Context;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

class MediaRecorder implements RecorderBase {
  private static final String LOG_TAG = "Record - MR";

  private boolean isRecording = false;
  private boolean isPaused = false;

  private android.media.MediaRecorder recorder = null;
  private String path;
  private Double maxAmplitude = -160.0;
  private final Context context;

  MediaRecorder(Context context) {
    this.context = context;
  }

  @Override
  @SuppressWarnings("deprecation")
  public void start(
      @NonNull String path,
      String encoder,
      int bitRate,
      int samplingRate,
      int numChannels,
      Map<String, Object> device
  ) throws Exception {
    stopRecording();

    this.path = path;

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
      recorder = new android.media.MediaRecorder(context);
    } else {
      recorder = new android.media.MediaRecorder();
    }

    // clamp channels
    numChannels = Math.max(1, numChannels);

    recorder.setAudioSource(android.media.MediaRecorder.AudioSource.DEFAULT);
    recorder.setAudioEncodingBitRate(bitRate);
    recorder.setAudioSamplingRate(samplingRate);
    recorder.setAudioChannels(numChannels);
    recorder.setOutputFormat(getOutputFormat(encoder));

    // must be set after output format
    Integer format = getEncoder(encoder);
    if (format == null) {
      Log.d(LOG_TAG, "Falling back to AAC LC");
      format = android.media.MediaRecorder.AudioEncoder.AAC;
    }

    recorder.setAudioEncoder(format);
    recorder.setOutputFile(path);

    try {
      recorder.prepare();
      recorder.start();
      isRecording = true;
      isPaused = false;
    } catch (IOException | IllegalStateException e) {
      recorder.release();
      recorder = null;
      throw new Exception(e);
    }
  }

  @Override
  public String stop() {
    stopRecording();
    return path;
  }

  @Override
  @RequiresApi(api = Build.VERSION_CODES.N)
  public void pause() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      pauseRecording();
    }
  }

  @Override
  @RequiresApi(api = Build.VERSION_CODES.N)
  public void resume() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      resumeRecording();
    }
  }

  @Override
  public boolean isRecording() {
    return isRecording;
  }

  @Override
  public boolean isPaused() {
    return isPaused;
  }

  @Override
  public Map<String, Object> getAmplitude() {
    Map<String, Object> amp = new HashMap<>();

    double current = -160.0;

    if (isRecording) {
      current = 20 * Math.log10(recorder.getMaxAmplitude() / 32768.0);

      if (current > maxAmplitude) {
        maxAmplitude = current;
      }
    }

    amp.put("current", current);
    amp.put("max", maxAmplitude);

    return amp;
  }

  @Override
  public boolean isEncoderSupported(String encoder) {
    Integer format = getEncoder(encoder);
    return format != null;
  }

  @Override
  public void close() {
    stopRecording();
  }

  private void stopRecording() {
    if (recorder != null) {
      try {
        if (isRecording || isPaused) {
          recorder.stop();
        }
      } catch (RuntimeException ex) {
        // Mute this exception since 'isRecording' can't be 100% sure
      } finally {
        recorder.reset();
        recorder.release();
        recorder = null;
      }
    }

    isRecording = false;
    isPaused = false;
    maxAmplitude = -160.0;
  }

  @RequiresApi(api = Build.VERSION_CODES.N)
  private void pauseRecording() {
    if (recorder != null) {
      try {
        if (isRecording) {
          recorder.pause();
          isPaused = true;
        }
      } catch (IllegalStateException ex) {
        Log.d(LOG_TAG, "Did you call pause() before before start() or after stop()?\n" + ex.getMessage());
      }
    }
  }

  @RequiresApi(api = Build.VERSION_CODES.N)
  private void resumeRecording() {
    if (recorder != null) {
      try {
        if (isPaused) {
          recorder.resume();
          isPaused = false;
        }
      } catch (IllegalStateException ex) {
        Log.d(LOG_TAG, "Did you call resume() before before start() or after stop()?\n" + ex.getMessage());
      }
    }
  }

  private int getOutputFormat(String encoder) {
    switch (encoder) {
      case "aacLc":
      case "aacEld":
      case "aacHe":
        return android.media.MediaRecorder.OutputFormat.MPEG_4;
      case "amrNb":
      case "amrWb":
        return android.media.MediaRecorder.OutputFormat.THREE_GPP;
      case "opus":
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
          return android.media.MediaRecorder.OutputFormat.OGG;
        }
      case "vorbisOgg":
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
          return android.media.MediaRecorder.OutputFormat.OGG;
        }
        return android.media.MediaRecorder.OutputFormat.MPEG_4;
    }

    return android.media.MediaRecorder.OutputFormat.DEFAULT;
  }

  // https://developer.android.com/reference/android/media/MediaRecorder.AudioEncoder
  private Integer getEncoder(String encoder) {
    switch (encoder) {
      case "aacLc":
        return android.media.MediaRecorder.AudioEncoder.AAC;
      case "aacEld":
        return android.media.MediaRecorder.AudioEncoder.AAC_ELD;
      case "aacHe":
        return android.media.MediaRecorder.AudioEncoder.HE_AAC;
      case "amrNb":
        return android.media.MediaRecorder.AudioEncoder.AMR_NB;
      case "amrWb":
        return android.media.MediaRecorder.AudioEncoder.AMR_WB;
      case "opus":
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
          return android.media.MediaRecorder.AudioEncoder.OPUS;
        }
      case "vorbisOgg":
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
          return android.media.MediaRecorder.AudioEncoder.VORBIS;
        }
    }

    return null;
  }
}

File: record_android/android/src/main/kotlin/com/llfbandit/record/record/recorder/MediaRecorder.kt
package com.llfbandit.record.record.recorder

import android.content.Context
import android.media.MediaRecorder
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import com.llfbandit.record.Utils
import com.llfbandit.record.record.AudioEncoder
import com.llfbandit.record.record.RecordConfig
import com.llfbandit.record.record.RecordState
import com.llfbandit.record.record.stream.RecorderStateStreamHandler
import java.io.IOException
import kotlin.math.log10

class MediaRecorder(
    private val context: Context,
    private val recorderStateStreamHandler: RecorderStateStreamHandler,
) : IRecorder {
    companion object {
        private val TAG = MediaRecorder::class.java.simpleName
    }

    // Signals whether a recording is in progress (true) or not (false).
    private var mIsRecording = false

    // Signals whether a recording is paused (true) or not (false).
    private var mIsPaused = false

    private var mRecorder: MediaRecorder? = null

    // Amplitude
    private var mMaxAmplitude = -160.0

    // Recording config
    private var mConfig: RecordConfig? = null

    @Throws(Exception::class)
    override fun start(config: RecordConfig) {
        stopRecording()

        val recorder = if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            @Suppress("DEPRECATION")
            MediaRecorder()
        } else {
            MediaRecorder(context)
        }

        recorder.setAudioSource(config.audioSource)
        recorder.setAudioEncodingBitRate(config.bitRate)
        recorder.setAudioSamplingRate(config.sampleRate)
        recorder.setAudioChannels(2.coerceAtMost(1.coerceAtLeast(config.numChannels)))
        recorder.setOutputFormat(getOutputFormat(config.encoder))
        // must be set after output format
        recorder.setAudioEncoder(getEncoder(config.encoder))
        recorder.setOutputFile(config.path)

        try {
            recorder.prepare()
            recorder.start()

            mConfig = config
            mRecorder = recorder

            updateState(RecordState.RECORD)
        } catch (e: IOException) {
            recorder.release()
            throw Exception(e)
        } catch (e: IllegalStateException) {
            recorder.release()
            throw Exception(e)
        }
    }

    override fun stop(stopCb: ((path: String?) -> Unit)?) {
        stopRecording()
        stopCb?.invoke(mConfig?.path)
    }

    override fun cancel() {
        stopRecording()
        Utils.deleteFile(mConfig?.path)
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    override fun pause() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            pauseRecording()
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    override fun resume() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            resumeRecording()
        }
    }

    override val isRecording: Boolean
        get() = mIsRecording
    override val isPaused: Boolean
        get() = mIsPaused

    override fun getAmplitude(): List<Double> {
        var current = -160.0

        if (mIsRecording) {
            current = 20 * log10(mRecorder!!.maxAmplitude / 32768.0)

            if (current > mMaxAmplitude) {
                mMaxAmplitude = current
            }
        }

        val amps: MutableList<Double> = ArrayList()
        amps.add(current)
        amps.add(mMaxAmplitude)
        return amps
    }

    override fun dispose() {
        stopRecording()
    }

    private fun stopRecording() {
        if (mRecorder != null) {
            try {
                if (mIsRecording || mIsPaused) {
                    mRecorder!!.stop()
                }
            } catch (ex: RuntimeException) {
                // Mute this exception since 'isRecording' can't be 100% sure
            } finally {
                mRecorder!!.reset()
                mRecorder!!.release()
                mRecorder = null
            }
        }

        updateState(RecordState.STOP)
        mMaxAmplitude = -160.0
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private fun pauseRecording() {
        if (mRecorder != null) {
            try {
                if (mIsRecording) {
                    mRecorder!!.pause()
                    updateState(RecordState.PAUSE)
                }
            } catch (ex: IllegalStateException) {
                Log.d(
                    TAG,
                    """
                        Did you call pause() before before start() or after stop()?
                        ${ex.message}
                        """.trimIndent()
                )
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private fun resumeRecording() {
        if (mRecorder != null) {
            try {
                if (mIsPaused) {
                    mRecorder!!.resume()
                    updateState(RecordState.RECORD)
                }
            } catch (ex: IllegalStateException) {
                Log.d(
                    TAG,
                    """
                        Did you call resume() before before start() or after stop()?
                        ${ex.message}
                        """.trimIndent()
                )
            }
        }
    }

    private fun updateState(state: RecordState) {
        when (state) {
            RecordState.PAUSE -> {
                mIsRecording = true
                mIsPaused = true
                recorderStateStreamHandler.sendStateEvent(RecordState.PAUSE.id)
            }

            RecordState.RECORD -> {
                mIsRecording = true
                mIsPaused = false
                recorderStateStreamHandler.sendStateEvent(RecordState.RECORD.id)
            }

            RecordState.STOP -> {
                mIsRecording = false
                mIsPaused = false
                recorderStateStreamHandler.sendStateEvent(RecordState.STOP.id)
            }
        }
    }

    private fun getOutputFormat(encoder: String): Int {
        return when (encoder) {
            AudioEncoder.aacLc, AudioEncoder.aacEld, AudioEncoder.aacHe -> MediaRecorder.OutputFormat.MPEG_4
            AudioEncoder.amrNb, AudioEncoder.amrWb -> MediaRecorder.OutputFormat.THREE_GPP
            AudioEncoder.opus -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    MediaRecorder.OutputFormat.OGG
                } else {
                    MediaRecorder.OutputFormat.MPEG_4
                }
            }

            else -> MediaRecorder.OutputFormat.DEFAULT
        }
    }

    // https://developer.android.com/reference/android/media/MediaRecorder.AudioEncoder
    private fun getEncoder(encoder: String): Int {
        return when (encoder) {
            AudioEncoder.aacLc -> MediaRecorder.AudioEncoder.AAC
            AudioEncoder.aacEld -> MediaRecorder.AudioEncoder.AAC_ELD
            AudioEncoder.aacHe -> MediaRecorder.AudioEncoder.HE_AAC
            AudioEncoder.amrNb -> MediaRecorder.AudioEncoder.AMR_NB
            AudioEncoder.amrWb -> MediaRecorder.AudioEncoder.AMR_WB
            AudioEncoder.opus -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    MediaRecorder.AudioEncoder.OPUS
                } else {
                    Log.d(TAG, "Falling back to AAC LC")
                    MediaRecorder.AudioEncoder.AAC
                }
            }

            else -> {
                Log.d(TAG, "Falling back to AAC LC")
                MediaRecorder.AudioEncoder.AAC
            }
        }
    }
}
